File: mycelium-py/src/mycelium/prompt_skill.py
"""
mycelium.prompt_skill — HTTP gateway client for LLM Prompt Skills.

Wraps the ``/gateway/prompts/*`` and ``/gateway/llm/*`` endpoints exposed
by a Rust Mycelium node compiled with ``--features llm``.

Example::

    import asyncio
    from mycelium.prompt_skill import PromptSkillClient, PromptTemplate

    async def main():
        client = PromptSkillClient("127.0.0.1", 7946)

        # Update a template already registered by a Rust node
        await client.update_prompt("ai", "chat", PromptTemplate(
            system="You are a helpful assistant.",
            user_template="{{input}}",
            max_tokens=512,
            temperature=0.7,
        ))

        # Call a skill
        result = await client.call("ai", "chat", "Hello!")
        print(result["output"])

        # List all visible skills
        for entry in await client.list():
            print(entry["ns"], entry["name"])

    asyncio.run(main())
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional

from ._pool import ClientPool


@dataclass
class PromptTemplate:
    """Mirror of the Rust ``PromptTemplate`` struct stored in cluster KV."""

    system: str
    user_template: str
    max_tokens: int = 512
    temperature: float = 0.7
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "system": self.system,
            "user_template": self.user_template,
            "max_tokens": self.max_tokens,
            "temperature": self.temperature,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "PromptTemplate":
        return cls(
            system=d.get("system", ""),
            user_template=d.get("user_template", ""),
            max_tokens=int(d.get("max_tokens", 512)),
            temperature=float(d.get("temperature", 0.7)),
            metadata=d.get("metadata", {}),
        )


class PromptSkillClient:
    """
    HTTP client for the Mycelium Prompt Skills gateway.

    Talks to ``/gateway/prompts/*`` and ``/gateway/llm/*`` on the Rust node.
    Requires the node to be compiled with ``--features llm``.

    :param host: Hostname or IP of the Mycelium HTTP gateway.
    :param port: HTTP port (default 8080).
    :param timeout: Default request timeout in seconds.
    """

    def __init__(
        self,
        host: str,
        port: int = 8080,
        timeout: float = 30.0,
    ) -> None:
        self._base = f"http://{host}:{port}"
        self._timeout = timeout
        # Pooled, loop-aware (see _pool.py): the previous eager AsyncClient was
        # bound to whichever loop first used it, breaking a handle reused
        # across separate asyncio.run() calls.
        self._pool = ClientPool(self._base, timeout)

    # ── Template management ────────────────────────────────────────────────────

    async def list(self) -> list[dict[str, Any]]:
        """
        List all prompt templates visible in the local KV snapshot.

        Returns a list of dicts with keys ``ns``, ``name``, ``max_tokens``,
        ``temperature``, ``metadata``.
        """
        async with self._pool.asy() as c:
            resp = await c.get("/gateway/prompts")
        resp.raise_for_status()
        result = resp.json()
        return result if isinstance(result, list) else []

    async def get(self, ns: str, name: str) -> Optional[PromptTemplate]:
        """
        Retrieve a specific prompt template from the local KV snapshot.

        Returns ``None`` if the key does not exist.
        """
        async with self._pool.asy() as c:
            resp = await c.get(f"/gateway/prompts/{ns}/{name}")
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        return PromptTemplate.from_dict(resp.json())

    async def update_prompt(self, ns: str, name: str, template: PromptTemplate) -> None:
        """
        Write (or overwrite) a prompt template in the cluster KV.

        The change propagates to all nodes via gossip. Serving nodes read the
        template fresh from KV on every invocation, so the update takes effect
        immediately without restarting any skill handler.
        """
        async with self._pool.asy() as c:
            resp = await c.put(
                f"/gateway/prompts/{ns}/{name}",
                json=template.to_dict(),
            )
        resp.raise_for_status()

    async def delete_prompt(self, ns: str, name: str) -> None:
        """
        Tombstone a prompt template in the cluster KV.

        The skill becomes unreachable once all serving nodes' capability entries
        expire (within 30 s). For a graceful drain, drop all ``PromptSkillHandle``
        objects on the Rust side first so capability entries evaporate naturally.
        """
        async with self._pool.asy() as c:
            resp = await c.delete(f"/gateway/prompts/{ns}/{name}")
        resp.raise_for_status()

    # ── Skill invocation ───────────────────────────────────────────────────────

    async def call(
        self,
        ns: str,
        name: str,
        input: str,
        context: Optional[dict[str, str]] = None,
        timeout_ms: int = 30_000,
    ) -> dict[str, Any]:
        """
        Invoke a prompt skill via the gateway.

        Resolves a provider for capability ``(ns, name)`` on the Rust node, sends
        an ``llm.invoke`` RPC, and returns the result dict::

            {"output": "...", "provider": "ip:port"}

        On error raises :class:`httpx.HTTPStatusError` — the gateway reports
        failures via status codes (404 no provider, 502 provider-side error,
        504 RPC timeout); the response body carries
        ``{"error": "...", "detail": "..."}``.

        :param ns:         Capability namespace (e.g. ``"ai"``).
        :param name:       Capability name (e.g. ``"chat"``).
        :param input:      The ``{{input}}`` value rendered into the template.
        :param context:    Optional extra ``{{variable}}`` substitutions.
        :param timeout_ms: RPC timeout in milliseconds.
        """
        body: dict[str, Any] = {
            "ns": ns,
            "name": name,
            "input": input,
            "context": context or {},
            "timeout_ms": timeout_ms,
        }
        async with self._pool.asy(timeout=timeout_ms / 1000.0 + 5.0) as c:
            resp = await c.post("/gateway/llm/call", json=body)
        resp.raise_for_status()
        return resp.json()

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._pool.aclose()

    async def __aenter__(self) -> "PromptSkillClient":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()

File: mycelium-ts/src/prompt_skill.ts
import { authHeaders, resolveToken, type AuthOptions } from "./auth";
/**
 * mycelium/prompt_skill — TypeScript client for LLM Prompt Skills.
 *
 * Wraps the `/gateway/prompts/*` and `/gateway/llm/*` endpoints exposed by a
 * Rust Mycelium node compiled with `--features llm`.
 *
 * @example
 * ```ts
 * import { PromptSkillClient, PromptTemplate } from "mycelium-ts";
 *
 * const client = new PromptSkillClient("127.0.0.1", 7946);
 *
 * await client.updatePrompt("ai", "chat", {
 *   system: "You are a helpful assistant.",
 *   userTemplate: "{{input}}",
 *   maxTokens: 512,
 *   temperature: 0.7,
 *   metadata: {},
 * });
 *
 * const result = await client.call("ai", "chat", "Hello!");
 * console.log(result.output);
 *
 * const entries = await client.list();
 * for (const e of entries) console.log(e.ns, e.name);
 *
 * await client.close();
 * ```
 */

/** Mirror of the Rust `PromptTemplate` struct stored in cluster KV. */
export interface PromptTemplate {
  /** System prompt. May contain `{{variable}}` placeholders. */
  system: string;
  /** User message template. Must contain at least `{{input}}`. */
  userTemplate: string;
  /** Maximum tokens in the LLM response. */
  maxTokens: number;
  /** Sampling temperature. `0.0` = deterministic. */
  temperature: number;
  /** Arbitrary metadata (tags, version notes, model hints). */
  metadata: Record<string, unknown>;
}

/** Result returned by {@link PromptSkillClient.call}. */
export interface CallResult {
  /** LLM-generated output text. */
  output: string;
  /** NodeId of the provider that served the request. */
  provider: string;
}

/** Summary entry returned by {@link PromptSkillClient.list}. */
export interface PromptListEntry {
  ns: string;
  name: string;
  maxTokens: number;
  temperature: number;
  metadata: Record<string, unknown>;
}

/** Wire shape sent to/from the Rust gateway (snake_case). */
interface TemplateWire {
  system: string;
  user_template: string;
  max_tokens: number;
  temperature: number;
  metadata: Record<string, unknown>;
}

interface ListEntryWire {
  ns: string;
  name: string;
  max_tokens: number;
  temperature: number;
  metadata: Record<string, unknown>;
}

function templateToWire(t: PromptTemplate): TemplateWire {
  return {
    system: t.system,
    user_template: t.userTemplate,
    max_tokens: t.maxTokens,
    temperature: t.temperature,
    metadata: t.metadata,
  };
}

function templateFromWire(w: TemplateWire): PromptTemplate {
  return {
    system: w.system,
    userTemplate: w.user_template,
    maxTokens: w.max_tokens,
    temperature: w.temperature,
    metadata: w.metadata,
  };
}

/**
 * HTTP client for the Mycelium Prompt Skills gateway.
 *
 * Talks to `/gateway/prompts/*` and `/gateway/llm/*` on the Rust node.
 * Requires the node to be compiled with `--features llm`.
 */
export class PromptSkillClient {
  private readonly baseUrl: string;
  private readonly auth: Record<string, string>;
  private readonly defaultTimeoutMs: number;

  constructor(
    host: string,
    port = 8080,
    timeoutMs = 30_000,
    opts: AuthOptions = {},
  ) {
    this.baseUrl = `http://${host}:${port}`;
    this.defaultTimeoutMs = timeoutMs;
    this.auth = authHeaders(resolveToken(opts.token));
  }

  // ── Template management ────────────────────────────────────────────────────

  /**
   * List all prompt templates visible in the local KV snapshot.
   *
   * Returns a summary list; use {@link get} to fetch the full template.
   */
  async list(): Promise<PromptListEntry[]> {
    const resp = await this.fetch("/gateway/prompts");
    const raw = (await resp.json()) as ListEntryWire[];
    return (Array.isArray(raw) ? raw : []).map((w) => ({
      ns:          w.ns,
      name:        w.name,
      maxTokens:   w.max_tokens,
      temperature: w.temperature,
      metadata:    w.metadata,
    }));
  }

  /**
   * Retrieve a specific prompt template from the local KV snapshot.
   *
   * Returns `null` if the key does not exist.
   */
  async get(ns: string, name: string): Promise<PromptTemplate | null> {
    const resp = await this.fetch(`/gateway/prompts/${ns}/${name}`, {
      ignoreNotFound: true,
    });
    if (resp.status === 404) return null;
    const w = (await resp.json()) as TemplateWire;
    return templateFromWire(w);
  }

  /**
   * Write (or overwrite) a prompt template in the cluster KV.
   *
   * The change propagates to all nodes via gossip. Serving nodes read the
   * template fresh from KV on every invocation, so the update takes effect
   * immediately without restarting any skill handler.
   */
  async updatePrompt(ns: string, name: string, template: PromptTemplate): Promise<void> {
    await this.fetch(`/gateway/prompts/${ns}/${name}`, {
      method: "PUT",
      body: JSON.stringify(templateToWire(template)),
      headers: { "Content-Type": "application/json" },
    });
  }

  /**
   * Tombstone a prompt template in the cluster KV.
   *
   * The skill becomes unreachable once all serving nodes' capability entries
   * expire (within 30 s).
   */
  async deletePrompt(ns: string, name: string): Promise<void> {
    await this.fetch(`/gateway/prompts/${ns}/${name}`, { method: "DELETE" });
  }

  // ── Skill invocation ───────────────────────────────────────────────────────

  /**
   * Invoke a prompt skill via the gateway.
   *
   * Resolves a provider for capability `(ns, name)`, sends an `llm.invoke` RPC,
   * and returns `{ output, provider }`.
   *
   * @param ns         Capability namespace (e.g. `"ai"`).
   * @param name       Capability name (e.g. `"chat"`).
   * @param input      The `{{input}}` value rendered into the template.
   * @param context    Optional extra `{{variable}}` substitutions.
   * @param timeoutMs  RPC timeout in milliseconds.
   */
  async call(
    ns: string,
    name: string,
    input: string,
    context: Record<string, string> = {},
    timeoutMs: number = this.defaultTimeoutMs,
  ): Promise<CallResult> {
    const resp = await this.fetch("/gateway/llm/call", {
      method: "POST",
      body: JSON.stringify({ ns, name, input, context, timeout_ms: timeoutMs }),
      headers: { "Content-Type": "application/json" },
    });
    const data = (await resp.json()) as {
      error?: string;
      detail?: string;
      output?: string;
      provider?: string;
    };
    if (data.error) {
      throw new Error(`llm call failed: ${data.error}: ${data.detail ?? ""}`);
    }
    return { output: String(data.output ?? ""), provider: String(data.provider ?? "") };
  }

  /** Close the client (no-op — fetch is stateless; kept for API symmetry). */
  close(): void { /* stateless fetch — nothing to close */ }

  // ── Internal helpers ───────────────────────────────────────────────────────

  private async fetch(
    path: string,
    opts: RequestInit & { ignoreNotFound?: boolean } = {},
  ): Promise<Response> {
    const { ignoreNotFound, headers, ...rest } = opts;
    const fetchOpts: RequestInit = {
      ...rest,
      headers: { ...this.auth, ...((headers as Record<string, string> | undefined) ?? {}) },
    };
    const resp = await globalThis.fetch(`${this.baseUrl}${path}`, fetchOpts);
    if (!resp.ok && !(ignoreNotFound && resp.status === 404)) {
      const text = await resp.text().catch(() => "");
      throw new Error(`Mycelium gateway error ${resp.status}: ${text}`);
    }
    return resp;
  }
}
